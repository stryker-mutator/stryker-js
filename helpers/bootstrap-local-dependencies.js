// @ts-check
import { LocalInstaller, progress } from 'install-local';
import fs from 'fs';
import glob from 'glob';
import path from 'path';

console.log('starting installation of local dependencies');

/**
 * @param {string} pattern 
 * @param {glob.IOptions} options 
 * @returns {Promise<string[]>}
 */
function globAsPromised(pattern, options) {
  return new Promise((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

/**
 * @typedef Package
 * @property {{ [name: string]: string }} localDependencies

/**
 * Installs local dependencies in one go,
 * reads the package.json files and installs the local dependencies marked there.
 * The packageJsonGlobs default to ./package.json and test/* /package.json files 
 * @param {string} directory
 * @param {string[]} packageJsonGlobs
 * @returns {Promise<void>}
 */
export async function bootstrapLocalDependencies(directory, packageJsonGlobs = ['package.json', 'test/*/package.json']) {
  console.log(`bootstrap ${path.resolve(directory)} (using ${packageJsonGlobs.join(',')})`);
  const files = (await Promise.all(packageJsonGlobs.map((globPattern) => globAsPromised(globPattern, { cwd: path.resolve(directory) })))).flat();
  const packages = await Promise.all(files.map(fileName => fs.promises.readFile(fileName, 'utf-8')
    .then((rawContent) => {
      /**
       * @type {Package}
       */
      const content = JSON.parse(rawContent);
      return { dir: path.dirname(fileName), content };
    })));
  /**
   * @type {import('install-local').ListByPackage}
   */
  const sourcesByTarget = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
