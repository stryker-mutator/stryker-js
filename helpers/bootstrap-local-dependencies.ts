import { ListByPackage, LocalInstaller, progress } from 'install-local';
import * as fs from 'fs';
import glob = require('glob');
import path = require('path');

console.log('starting installation of local dependencies');

function globAsPromised(pattern: string, options: glob.IOptions) {
  return new Promise<string[]>((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

function readFile(fileName: string) {
  return new Promise<string>((res, rej) => fs.readFile(fileName, 'utf8', (err, content) => {
    if (err) {
      rej(err);
    } else {
      res(content);
    }
  }));
}

interface Package {
  localDependencies?: { [name: string]: string };
}

/**
 * Installs local dependencies in one go,
 * reads package.json and test/* /package.json files and installs the local dependencies marked there
 * @param directory the directory where the tests live
 */
export async function bootstrapLocalDependencies(directory: string) {
  console.log('bootstrap ' + path.resolve(directory));
  const files = await globAsPromised('{package.json,test/*/package.json}', { cwd: path.resolve(directory) });
  const packages = await Promise.all(files.map(fileName => readFile(fileName)
    .then(content => ({ dir: path.dirname(fileName), content: JSON.parse(content) as Package }))));
  const sourcesByTarget: ListByPackage = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
