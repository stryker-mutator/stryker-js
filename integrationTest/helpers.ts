import * as path from 'path';
import * as fs from 'mz/fs';
import { expect } from 'chai';
import { ScoreResult } from 'stryker-api/report';

export async function readScoreResult(eventResultDirectory = path.resolve('reports', 'mutation', 'events')) {
    const allReportFiles = await fs.readdir(eventResultDirectory);
    const scoreResultReportFile = allReportFiles.find(file => !!file.match(/.*onScoreCalculated.*/));
    expect(scoreResultReportFile).ok;
    const scoreResultContent = await fs.readFile(path.resolve(eventResultDirectory, scoreResultReportFile || ''), 'utf8');
    return JSON.parse(scoreResultContent) as ScoreResult;
}