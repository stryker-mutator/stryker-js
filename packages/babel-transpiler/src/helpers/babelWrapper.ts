/**
 * This wrapper is needed because babel/core only exports
 * non-stubbable es6 properties. See node_modules/@babel/core/lib/index.js
 */
export * from '@babel/core';
