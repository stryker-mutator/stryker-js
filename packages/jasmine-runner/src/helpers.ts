import { TestResult, TestStatus } from '@stryker-mutator/api/test-runner';
import jasmine from 'jasmine';

export const helpers = {
  toStrykerTestResult(specResult: jasmine.SpecResult, timeSpentMs: number): TestResult {
    const baseResult = {
      id: specResult.id.toString(),
      name: specResult.fullName,
      timeSpentMs,
    };
    if (specResult.status === 'disabled' || specResult.status === 'pending' || specResult.status === 'excluded') {
      return {
        ...baseResult,
        status: TestStatus.Skipped,
      };
    } else if (!specResult.failedExpectations || specResult.failedExpectations.length === 0) {
      return {
        ...baseResult,
        status: TestStatus.Success,
      };
    } else {
      return {
        ...baseResult,
        status: TestStatus.Failed,
        failureMessage: specResult.failedExpectations.map((failedExpectation) => failedExpectation.message).join(','),
      };
    }
  },
  createJasmine(options: jasmine.JasmineOptions): jasmine {
    return new jasmine(options);
  },
};
