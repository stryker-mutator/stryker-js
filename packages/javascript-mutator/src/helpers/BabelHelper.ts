import generate from '@babel/generator';
import { parse, ParserOptions } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import * as types from '@babel/types';
import { NodeWithParent } from './ParentNode';

export default class BabelHelper {
  public static parse(code: string): types.File {
    return parse(code, this.createOptions());
  }

  private static createOptions(): ParserOptions {
    return {
      plugins: [
        'asyncGenerators',
        'bigInt',
        'classProperties',
        'dynamicImport',
        'flow',
        'jsx',
        'objectRestSpread',
        ['decorators', { decoratorsBeforeExport: true }] as any
      ],
      sourceType: 'unambiguous'
    };
  }

  public static getNodes(ast: types.File): NodeWithParent[] {
    const nodes: NodeWithParent[] = [];

    traverse(ast, {
      enter(path: NodePath<types.Node>) {
        const node: NodeWithParent = path.node;
        node.parent = path.parent as any;
        Object.freeze(node);
        nodes.push(node);
      }
    });

    return nodes;
  }

  public static generateCode(ast: types.Node) {
    return generate(ast).code;
  }
}
