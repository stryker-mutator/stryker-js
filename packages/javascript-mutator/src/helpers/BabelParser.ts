import generate from '@babel/generator';
import { parse, ParserOptions } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import * as types from '@babel/types';
import { NodeWithParent } from './ParentNode';
import { Logger } from '@stryker-mutator/api/logging';
import { commonTokens, tokens } from '@stryker-mutator/api/plugin';
import { MutatorDescriptor } from '@stryker-mutator/api/core';

export default class BabelParser {
  public static inject = tokens(commonTokens.logger, commonTokens.mutatorDescriptor);
  private readonly options: ParserOptions;

  constructor(private readonly log: Logger, mutatorDescriptor: MutatorDescriptor) {
    this.options = this.createOptions(mutatorDescriptor.plugins);
  }

  public parse(code: string): types.File {
    return parse(code, this.options);
  }

  private createOptions(pluginOverrides: string[] | null): ParserOptions {
    const plugins = pluginOverrides || [
      'asyncGenerators',
      'bigInt',
      'classProperties',
      'dynamicImport',
      'flow',
      'jsx',
      'objectRestSpread',
      ['decorators', { decoratorsBeforeExport: true }] as any
    ];
    const options: ParserOptions = {
      plugins,
      sourceType: 'unambiguous'
    };
    if (this.log.isDebugEnabled()) {
      this.log.debug(`Using options ${JSON.stringify(options)}`);
    }
    return options;
  }

  public getNodes(ast: types.File): NodeWithParent[] {
    const nodes: NodeWithParent[] = [];

    traverse(ast, {
      enter(path: NodePath<types.Node>) {
        const node: NodeWithParent = path.node;
        node.parent = path.parent as any;
        Object.freeze(node);
        nodes.push(node);
      }
    });

    return nodes;
  }

  public generateCode(ast: types.Node) {
    return generate(ast).code;
  }
}
