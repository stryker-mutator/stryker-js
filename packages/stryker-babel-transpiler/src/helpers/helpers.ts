import * as path from 'path';

export function toJSFileName(fileName: string) {
  const fileNameInfo = path.parse(fileName);
  const newFileName = path.join(fileNameInfo.dir, fileNameInfo.name + '.js');
  return newFileName;
}
