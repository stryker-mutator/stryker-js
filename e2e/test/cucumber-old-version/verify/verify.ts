import { expectMetricsJson } from '../../../helpers';

describe('After running stryker on a cucumber-ts project', () => {
  it('should report expected scores', async () => {
    await expectMetricsJson({
      killed: 4,
      ignored: 0,
      survived: 0,
      noCoverage: 2,
      runtimeErrors: 2,
      mutationScore: 66.67,
    });
    /*
----------------------|---------|----------|-----------|------------|----------|---------|
File                  | % score | # killed | # timeout | # survived | # no cov | # error |
----------------------|---------|----------|-----------|------------|----------|---------|
All files             |   66.67 |        4 |         0 |          0 |        2 |       2 |*/
  });
});
