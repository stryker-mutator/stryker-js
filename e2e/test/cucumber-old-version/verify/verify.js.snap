// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`After running stryker on a cucumber-ts project should report expected scores 1`] = `
Object {
  "compileErrors": 0,
  "ignored": 0,
  "killed": 4,
  "mutationScore": 66.66666666666666,
  "mutationScoreBasedOnCoveredCode": 100,
  "noCoverage": 2,
  "runtimeErrors": 2,
  "survived": 0,
  "timeout": 0,
  "totalCovered": 4,
  "totalDetected": 4,
  "totalInvalid": 2,
  "totalMutants": 8,
  "totalUndetected": 2,
  "totalValid": 6,
}
`;
